service/src/main/java/uk/nhs/adaptors/gp2gp/gpc/GpcRequestBuilder.java

```diff
@@ -52,6 +53,7 @@
     private static final int NUMBER_OF_RECENT_CONSULTANTS = 3;
     private static final String GPC_STRUCTURED_INTERACTION_ID = "urn:nhs:names:services:gpconnect:fhir:operation:gpc"
         + ".getstructuredrecord-1";
+    private static final String GPC_DOCUMENT_INTERACTION_ID = "urn:nhs:names:services:gpconnect:documents:fhir:rest:read:binary-1";
```

Comment:
What do you think about changing this class to implement Builder pattern with the help of some methods to configure the builder?

Implementing pattern I mean by something like that
```

   public RequestHeadersSpec<?> getStructuredRecordRequest(
        Parameters requestBodyParameters,
        GetGpcStructuredTaskDefinition structuredTaskDefinition
    ) {

        return new RequestBuilder(
                gpcTokenBuilder,
                gpcConfiguration,
                gpcWebClientFilter,
                HttpMethod.POST,
                gpcConfiguration.getStructuredEndpoint())
            .withInteractionId(GPC_STRUCTURED_INTERACTION_ID)
            .withTaskDefinition(structuredTaskDefinition)
            .withBody(fhirParser.encodeResourceToString(requestBodyParameters))
            .build();
    }

    public RequestHeadersSpec<?> getDocumentRecordRequest(GetGpcDocumentTaskDefinition documentTaskDefinition) {

        return new RequestBuilder(
                    gpcTokenBuilder,
                    gpcConfiguration,
                    gpcWebClientFilter,
                    HttpMethod.GET,
                    gpcConfiguration.getDocumentEndpoint() + "/" + documentTaskDefinition.getDocumentId())
            .withInteractionId(GPC_DOCUMENT_INTERACTION_ID)
            .withTaskDefinition(documentTaskDefinition)
            .build();
    }

    public static class RequestBuilder {

        private GpcTokenBuilder gpcTokenBuilder;
        private final GpcConfiguration gpcConfiguration;
        private final GpcWebClientFilter gpcWebClientFilter;

        private RequestBodySpec request;

        public RequestBuilder(
            GpcTokenBuilder gpcTokenBuilder,
            GpcConfiguration gpcConfiguration,
            GpcWebClientFilter gpcWebClientFilter,
            HttpMethod post,
            String url
        ) {

            this.gpcTokenBuilder = gpcTokenBuilder;
            this.gpcConfiguration = gpcConfiguration;
            this.gpcWebClientFilter = gpcWebClientFilter;

            SslContext sslContext = getSSLContext();
            HttpClient httpClient = HttpClient.create().secure(t -> t.sslContext(sslContext));
            WebClient client = getWebClient(httpClient);

            request = client
                .method(post)
                .uri(url);
        }

        public RequestBuilder withInteractionId(String interactionId) {
            this.request.header(SSP_INTERACTION_ID, interactionId);

            return this;
        }

        private RequestBuilder withTaskDefinition(TaskDefinition taskDefinition) {
            this.request
                .header(SSP_FROM, taskDefinition.getFromAsid())
                .header(SSP_TO, taskDefinition.getToAsid())
                .header(SSP_TRACE_ID, taskDefinition.getConversationId())
                .header(TARGET_HOST, gpcConfiguration.getHost())
                .header(AUTHORIZATION, AUTHORIZATION_BEARER + gpcTokenBuilder.buildToken(taskDefinition.getFromOdsCode()));

            return this;
        }

        public RequestBuilder withBody(String requestBody) {
            BodyInserter<Object, ReactiveHttpOutputMessage> bodyInserter
                = BodyInserters.fromValue(requestBody);

            this.request.body(bodyInserter)
                .header(CONTENT_LEN, valueOf(requestBody.length()));

            return this;
        }

        public RequestHeadersSpec<?> build() {
           return this.request
                .accept(MediaType.valueOf(FHIR_CONTENT_TYPE))
                .header(CONTENT_TYPE, FHIR_CONTENT_TYPE)
                .header(TARGET_HOST, gpcConfiguration.getHost());
        }

        @SneakyThrows
        private SslContext getSSLContext() {
            return SslContextBuilder
                .forClient()
                .trustManager(InsecureTrustManagerFactory.INSTANCE)
                .build();
        }

        private WebClient getWebClient(HttpClient httpClient) {
            return WebClient
                .builder()
                .exchangeStrategies(getExchangeStrategies())
                .clientConnector(new ReactorClientHttpConnector(httpClient))
                .filter(gpcWebClientFilter.errorHandlingFilter())
                .baseUrl(gpcConfiguration.getUrl())
                .defaultUriVariables(Collections.singletonMap("url", gpcConfiguration.getUrl()))
                .build();
        }

        private ExchangeStrategies getExchangeStrategies() {
            return ExchangeStrategies
                .builder()
                .codecs(
                    configurer -> configurer.defaultCodecs()
                        .maxInMemorySize(BYTE_COUNT)).build();
        }

    }



```